apps/ejabberd/src/mod_offline_riak.erl

```diff
@@ -0,0 +1,187 @@
+
+-spec pop_messages(LUser, LServer) -> {ok, Result} | {error, Reason} when
+    LUser :: ejabberd:luser(),
+    LServer :: ejabberd:lserver(),
+    Reason :: term(),
+    Result :: list(#offline_msg{}).
+pop_messages(LUser, LServer) ->
+    Keys = read_user_idx(LUser, LServer),
+    To = jid:make(LUser, LServer, <<>>),
+    Msgs = [pop_msg(Key, LUser, LServer, To) || Key <- Keys],
+    {ok, lists:flatten(Msgs)}.
+
+-spec write_messages(LUser, LServer, Msgs) ->
+    ok  when
+    LUser :: ejabberd:luser(),
+    LServer :: ejabberd:lserver(),
+    Msgs :: list().
+write_messages(LUser, LServer, Msgs) ->
+    [write_msg(LUser, LServer, Msg) || Msg <- Msgs],
+    ok.
+
+-spec count_offline_messages(LUser, LServer, MaxArchivedMsgs) ->
+    integer() when
+      LUser :: ejabberd:luser(),
+      LServer :: ejabberd:lserver(),
+      MaxArchivedMsgs :: integer().
+count_offline_messages(LUser, LServer, MaxArchivedMsgs) ->
+    {ok, IdxResult} = mongoose_riak:get_index(bucket_type(LServer), ?USER_IDX,
+                                              LUser, [{max_results, MaxArchivedMsgs}]),
+    length(IdxResult?INDEX_RESULTS.keys).
+
+-spec remove_expired_messages(Host) -> {error, Reason} | {ok, Count} when
+    Host :: ejabberd:lserver(),
+    Reason :: term(),
+    Count :: integer().
+remove_expired_messages(Host) ->
+    TimestampInt = usec:from_now(os:timestamp()),
+    {ok, Result} = mongoose_riak:get_index_range(bucket_type(Host), ?EXPIRE_IDX,
+                                                 0, TimestampInt, []),
+    Keys = Result?INDEX_RESULTS.keys,
+    [mongoose_riak:delete(bucket_type(Host), Key) || Key <- Keys],
+    {ok, length(Keys)}.
+
+-spec remove_old_messages(Host, Days) -> {error, Reason} | {ok, Count} when
+    Host :: ejabberd:lserver(),
+    Days :: erlang:timestamp(),
+    Reason :: term(),
+    Count :: integer().
+remove_old_messages(Host, Timestamp) ->
+    TimestampInt = usec:from_now(Timestamp),
+    {ok, Result} = mongoose_riak:get_index_range(bucket_type(Host), ?TIMESTAMP_IDX,
+                                                 0, TimestampInt, []),
+    Keys = Result?INDEX_RESULTS.keys,
+    [mongoose_riak:delete(bucket_type(Host), Key) || Key <- Keys],
+    {ok, length(Keys)}.
+
+-spec remove_user(LUser, LServer) -> any() when
+    LUser :: binary(),
+    LServer :: binary().
+
+remove_user(LUser, LServer) ->
+    Keys = read_user_idx(LUser, LServer),
+    [mongoose_riak:delete(bucket_type(LServer), Key) || Key <- Keys].
+
+read_user_idx(LUser, LServer) ->
+    {ok, IdxResult} = mongoose_riak:get_index(bucket_type(LServer), ?USER_IDX,
+                                              LUser, []),
+    IdxResult?INDEX_RESULTS.keys.
+
+write_msg(LUser, LServer, #offline_msg{from = FromJID, packet = Packet,
+                                       timestamp = TimestampIn, expire = Expire}) ->
+    Timestamp = usec:from_now(TimestampIn),
+    Obj = riakc_obj:new(bucket_type(LServer), key(LUser, Timestamp), exml:to_binary(Packet)),
```

Comment:
Actually they do. I've seen many questions like  "how to read this from other lang than Erlang"